File: dart/lib/otp_crypto/decryptor.dart
/// Decryptor – Verifies & decrypts a SecureMessage (Encrypt-then-MAC)
/// ------------------------------------------------------------------
/// Processing order (DO NOT change):
///   1) Validate protocol version and time-window skew.
///   2) Derive {encKey, macKey} via HKDF-SHA256 from config.
///   3) Recompute tag = HMAC(macKey, "tag" || u64be(w) || nonce || ciphertext).
///   4) Constant-time compare with body tag; if mismatch → AuthenticationFailed.
///   5) Derive IV = HMAC(macKey, "iv" || u64be(w) || nonce)[:16].
///   6) Decrypt AES-256-CBC + PKCS#7 using encKey+IV → plaintext.
///
/// NOTES:
/// - This class **does not** handle HTTP. It only consumes a `SecureMessage`
///   reconstructed from wire headers/body.
/// - Time skew tolerance is enforced **before** any crypto to fail fast.
/// - We always verify MAC **before** decrypting (Encrypt-then-MAC).
///
/// HINTS:
/// - Keep a single `Decryptor` around; it caches HKDF-derived keys.
/// - Configure `verificationSkewWindows` in `OtpCryptoConfig` to accept ±N
///   adjacent windows relative to *current* window.

import 'dart:typed_data';

import '../models/secure_message.dart';
import 'errors.dart';
import 'hkdf.dart';
import 'iv_deriver.dart';
import 'otp_cipher.dart';
import 'otp_crypto_config.dart';
import 'tag_deriver.dart';
import 'utils.dart';

class Decryptor {
  /// Global/effective config.
  final OtpCryptoConfig _cfg;

  /// Cached HKDF-derived keys (enc + mac).
  final DerivedKeys _keys;

  /// Creates a Decryptor bound to [OtpCryptoConfig].
  ///
  /// [config] Defaults to the global singleton `OtpCryptoConfig.instance`.
  ///
  /// HINT: Instantiate once and reuse; keys are derived in the constructor.
  Decryptor({OtpCryptoConfig? config})
      : _cfg = config ?? OtpCryptoConfig.instance,
        _keys = HkdfSha256.deriveKeys(
          masterKey: (config ?? OtpCryptoConfig.instance).masterKey,
          salt: (config ?? OtpCryptoConfig.instance).hkdfSalt,
          info: (config ?? OtpCryptoConfig.instance).hkdfInfo,
        );

  /// Verifies and decrypts a previously parsed `SecureMessage`.
  ///
  /// INPUT:
  /// - [msg]: a format-validated message (see `SecureMessage.fromWire`).
  ///
  /// OUTPUT:
  /// - plaintext bytes if authentication and decryption succeed.
  ///
  /// Throws:
  /// - [InvalidMessageException] for version mismatch or window out-of-range.
  /// - [AuthenticationFailedException] on HMAC mismatch.
  /// - [DecryptionFailedException] if AES decryption/padding fails.
  ///
  /// HINT:
  /// - Always call this on the server for incoming requests and on the client
  ///   for responses returned by the server.
  Uint8List unprotect(SecureMessage msg) {
    // 1) Protocol version check (extensible: allow only exact match for v1).
    if (msg.version != _cfg.protocolVersion) {
      throw InvalidMessageException(
        cause: ArgumentError.value(
          msg.version,
          'version',
          'Unsupported protocol version; expected ${_cfg.protocolVersion}.',
        ),
      );
    }

    // 2) Window skew check (reject too old/future messages quickly).
    _enforceWindowSkew(msg.window);

    // 3) Verify tag in constant time (Encrypt-then-MAC).
    final computedTag = TagDeriver.derive(
      macKey: _keys.macKey,
      window: msg.window,
      nonce: msg.nonce,
      ciphertext: msg.ciphertext,
    );
    final ok = Bytes.constantTimeEquals(computedTag, msg.tag);
    // Best-effort zeroization (mutable copy).
    Bytes.secureZero(computedTag);

    if (!ok) {
      throw AuthenticationFailedException();
    }

    // 4) Derive IV and decrypt.
    final iv = IvDeriver.derive(
      macKey: _keys.macKey,
      window: msg.window,
      nonce: msg.nonce,
    );

    // If AES fails (e.g., bad padding), OtpCipher throws DecryptionFailedException.
    final plaintext = OtpCipher.decrypt(
      encKey: _keys.encKey,
      iv: iv,
      ciphertext: msg.ciphertext,
    );

    return plaintext;
  }

  // -- helpers ---------------------------------------------------------------

  /// Enforces skew tolerance: abs(receivedW - currentW) <= verificationSkewWindows
  ///
  /// Throws [WindowOutOfRangeException] if outside tolerance.
  ///
  /// HINT: This prevents accepting stale/future messages even if tags match.
  void _enforceWindowSkew(int receivedW) {
    final currentW = _cfg.currentWindow();
    final delta = (receivedW - currentW).abs();
    if (delta > _cfg.verificationSkewWindows) {
      throw WindowOutOfRangeException(
        cause: StateError(
          'received window $receivedW outside tolerance of current $currentW (±${_cfg.verificationSkewWindows})',
        ),
      );
    }
  }
}

File: php/src/Http/Decryptor.php
<?php

declare(strict_types=1);

namespace OtpCrypto\Http;

use OtpCrypto\Crypto\OtpCryptoConfig;
use OtpCrypto\Crypto\Hkdf;
use OtpCrypto\Crypto\DerivedKeys;
use OtpCrypto\Crypto\TagDeriver;
use OtpCrypto\Crypto\IvDeriver;
use OtpCrypto\Crypto\OtpCipher;
use OtpCrypto\Crypto\Utils;
use OtpCrypto\Crypto\InvalidMessageException;
use OtpCrypto\Crypto\AuthenticationFailedException;
use OtpCrypto\Crypto\WindowOutOfRangeException;
use OtpCrypto\Crypto\DecryptionFailedException;
use OtpCrypto\Models\SecureMessage;

/**
 * Decryptor – Verifies & decrypts a SecureMessage (Encrypt-then-MAC)
 * ------------------------------------------------------------------
 * Processing order (DO NOT change):
 *   1) Validate protocol version and time-window skew.
 *   2) Derive {encKey, macKey} via HKDF-SHA256 from config (cached).
 *   3) Recompute tag = HMAC(macKey, "tag" || u64be(w) || nonce || ciphertext).
 *   4) Constant-time compare with body tag; if mismatch → AuthenticationFailed.
 *   5) Derive IV = HMAC(macKey, "iv" || u64be(w) || nonce)[:16].
 *   6) Decrypt AES-256-CBC + PKCS#7 using encKey+IV → plaintext.
 *
 * NOTES:
 * - This class does **not** do any HTTP. It only consumes a `SecureMessage`
 *   reconstructed from headers/body.
 * - Always verify the MAC **before** decryption (Encrypt-then-MAC).
 */
final class Decryptor
{
    /** Effective configuration (singleton by default). */
    private OtpCryptoConfig $cfg;

    /** Cached HKDF-derived keys (enc + mac). */
    private DerivedKeys $keys;

    /**
     * Creates a Decryptor bound to a given configuration.
     *
     * @param OtpCryptoConfig|null $config If null, uses OtpCryptoConfig::instance().
     *
     * HINT: Instantiate once and reuse; keys are derived in the constructor.
     */
    public function __construct(?OtpCryptoConfig $config = null)
    {
        $this->cfg = $config ?? OtpCryptoConfig::instance();
        $this->keys = Hkdf::deriveKeys(
            $this->cfg->masterKey(),
            $this->cfg->hkdfSalt(),
            $this->cfg->hkdfInfo()
        );
    }

    /**
     * Verifies and decrypts a previously parsed `SecureMessage`.
     *
     * INPUT:
     * - $msg: a format-validated message (see SecureMessage::fromWire).
     *
     * OUTPUT:
     * - plaintext bytes (binary string) if authentication and decryption succeed.
     *
     * @throws InvalidMessageException        on version mismatch.
     * @throws WindowOutOfRangeException      when outside the accepted time window.
     * @throws AuthenticationFailedException  on HMAC mismatch.
     * @throws DecryptionFailedException      if AES decryption/padding fails.
     */
    public function unprotect(SecureMessage $msg): string
    {
        // 1) Protocol version check (exact match for v1).
        if ($msg->v() !== $this->cfg->protocolVersion()) {
            throw new InvalidMessageException(
                new \InvalidArgumentException(
                    sprintf('Unsupported protocol version %d (expected %d)', $msg->v(), $this->cfg->protocolVersion())
                )
            );
        }

        // 2) Window skew check (reject too old/future messages quickly).
        $this->enforceWindowSkew($msg->w());

        // 3) Verify tag (Encrypt-then-MAC).
        $computedTag = TagDeriver::derive(
            $this->keys->macKey(),
            $msg->w(),
            $msg->nonce(),
            $msg->ciphertext()
        );

        if (!Utils::constantTimeEquals($computedTag, $msg->tag())) {
            throw new AuthenticationFailedException();
        }

        // 4) Derive IV and decrypt.
        $iv = IvDeriver::derive(
            $this->keys->macKey(),
            $msg->w(),
            $msg->nonce()
        );

        try {
            return OtpCipher::decrypt(
                $this->keys->encKey(),
                $iv,
                $msg->ciphertext()
            );
        } catch (\RuntimeException $e) {
            // Normalize OpenSSL errors to a generic, non-leaking exception.
            throw new DecryptionFailedException($e);
        }
    }

    // -- helpers --------------------------------------------------------------

    /**
     * Enforces skew tolerance: abs(receivedW - currentW) <= verificationSkewWindows
     *
     * @throws WindowOutOfRangeException if outside tolerance.
     */
    private function enforceWindowSkew(int $receivedW): void
    {
        $currentW = $this->cfg->currentWindow();
        $delta = abs($receivedW - $currentW);
        if ($delta > $this->cfg->verificationSkewWindows()) {
            throw new WindowOutOfRangeException(
                new \RuntimeException(
                    sprintf(
                        'received window %d outside tolerance of current %d (±%d)',
                        $receivedW,
                        $currentW,
                        $this->cfg->verificationSkewWindows()
                    )
                )
            );
        }
    }
}
